In [1]:
import pandas as pd

employees = pd.read_csv('data/omni_employees.csv', parse_dates=['start_date'])
pay_details = pd.read_csv('data/omni_pay_details.csv')
teams = pd.read_csv('data/omni_teams.csv')
committees = pd.read_csv('data/omni_committees.csv')
employees_committees = pd.read_csv('data/omni_employees_committees.csv')

### Data type for each column

In [2]:
employees.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   id             1000 non-null   int64         
 1   first_name     974 non-null    object        
 2   last_name      1000 non-null   object        
 3   email          878 non-null    object        
 4   department     1000 non-null   object        
 5   team_id        1000 non-null   int64         
 6   grade          980 non-null    float64       
 7   country        1000 non-null   object        
 8   fte_hours      1000 non-null   float64       
 9   pension_enrol  958 non-null    object        
 10  salary         935 non-null    float64       
 11  pay_detail_id  1000 non-null   int64         
 12  start_date     926 non-null    datetime64[ns]
dtypes: datetime64[ns](1), float64(3), int64(3), object(6)
memory usage: 101.7+ KB


### Max and min `salaray`

In [3]:
employees.describe()

id     team_id       grade    fte_hours        salary  \
count  1000.000000  1000.00000  980.000000  1000.000000    935.000000   
mean    500.500000     5.39900    0.189796     0.627750  59929.568984   
std     288.819436     2.88342    0.392340     0.278514  22891.672558   
min       1.000000     1.00000    0.000000     0.250000  20063.000000   
25%     250.750000     3.00000    0.000000     0.500000  40103.500000   
50%     500.500000     5.00000    0.000000     0.750000  59917.000000   
75%     750.250000     8.00000    0.000000     0.750000  78917.500000   
max    1000.000000    10.00000    1.000000     1.000000  99889.000000   

       pay_detail_id  
count    1000.000000  
mean      500.500000  
std       288.819436  
min         1.000000  
25%       250.750000  
50%       500.500000  
75%       750.250000  
max      1000.000000

Salary  
Max = 99889  
Min = 20063

### How many rows are there in the DataFrame?

In [4]:
employees.shape

(1000, 13)

1000 rows

### Unique values for `department` and `country`

In [5]:
employees['department'].nunique()

12

In [6]:
employees['country'].nunique()

130

### Q2. Find all the details of employees who work in the 'Legal' department

In [7]:
employees.loc[employees["department"] == "Legal"]

id first_name    last_name                           email department  \
0      1      Ibbie     Roscrigg            iroscrigg0@google.fr      Legal   
2      4     Osmund       Kittel          okittel3@bloomberg.com      Legal   
10    12  Thorstein         Garr                  tgarrb@icio.us      Legal   
17    19      Robby     Harragin                             NaN      Legal   
35    37     Franky        Idell          fidell10@economist.com      Legal   
..   ...        ...          ...                             ...        ...   
942  944     Velvet     Mellodey  vmellodeyq7@huffingtonpost.com      Legal   
958  960      Joela  McClenaghan    jmcclenaghanqn@aboutads.info      Legal   
965  967   Elsinore        Stein                             NaN      Legal   
980  982    Maurita      Sirkett             msirkettr9@webs.com      Legal   
992  994       Dani    Berthelet             dbertheletrl@si.edu      Legal   

     team_id  grade         country  fte_hours pension_enrol   salary  \
0          9    0.0         Nigeria       0.25           Yes  97667.0   
2         10    0.0  United Kingdom       1.00           Yes  51200.0   
10         1    0.0           China       0.75            No  39926.0   
17        10    0.0     South Korea       0.25           Yes  70830.0   
35         4    0.0          Sweden       1.00           Yes      NaN   
..       ...    ...             ...        ...           ...      ...   
942        2    0.0     Philippines       0.75           Yes  70961.0   
958       10    0.0           China       1.00           Yes  56952.0   
965        2    0.0        Malaysia       1.00           Yes      NaN   
980        6    0.0         Ecuador       0.25            No  97989.0   
992        5    1.0           Chile       1.00           Yes  85316.0   

     pay_detail_id start_date  
0                1 2014-12-25  
2                4 2007-09-06  
10              12 2012-03-23  
17              19 1998-07-20  
35              37 1990-06-06  
..             ...        ...  
942            944        NaT  
958            960 2011-02-19  
965            967 2013-01-25  
980            982 2006-12-09  
992            994 1999-09-24  

[102 rows x 13 columns]

### Q3. How many employees are based in Japan?

In [8]:
# Running .count() on the id column might help here, or you could use the .shape attribute

employees.loc[employees["country"] == "Japan"].id.count()

26

### Q4. [Harder] In the question above, we suggested you .count() the id column (i.e. treating it like a SQL primary key). But we haven't yet shown that id satisfies the associated requirements. Confirm that the number of unique values in id equals the number of rows in employees.

In [9]:
employees.shape

(1000, 13)

In [10]:
employees.id.nunique()

1000

In [11]:
employees.shape[0] == employees.id.nunique()

True

### Q5. How many employees have a missing email address?

In [12]:
employees.email.isna().sum()

122

### Q6. Calculate the mean salary of employees in the 'Legal' department.

In [13]:
employees.loc[employees["department"] == "Legal"].salary.mean()

56503.947916666664

### Q7. Obtain the first_name, last_name and salary of all employees sorted in descending order of salary.

In [14]:
employees[["first_name", "last_name", "salary"]].sort_values("salary",ascending=False)

first_name  last_name   salary
758     Gustave    Truwert  99889.0
945     Patrice     Chitty  99853.0
857       Corny      Yearn  99798.0
963      Brucie   Ceschini  99634.0
198     Katinka     Peffer  99565.0
..          ...        ...      ...
922       Claus   Hadigate      NaN
940      Isobel   McMillan      NaN
953    Martainn  McCaughan      NaN
965    Elsinore      Stein      NaN
983  Bernardine    Brasier      NaN

[1000 rows x 3 columns]

### Q8. Obtain the first_name, last_name and country of employees ordered first alphabetically by country and then alphabetically by last_name.

In [15]:
employees[["first_name", "last_name", "country"]].sort_values(["country", "last_name"])

first_name  last_name      country
929       Abeu     Pawden  Afghanistan
176      Trixi  Pickvance  Afghanistan
255      Vance     Ratlee  Afghanistan
792      Beale    Raynard  Afghanistan
338    Bentlee        Toy  Afghanistan
..         ...        ...          ...
691   Ebenezer     Roseby        Yemen
421     Gretta     Zealey        Yemen
358     Farrel    Clethro       Zambia
920       Conn     Robiot       Zambia
623      Aaron       Tack       Zambia

[1000 rows x 3 columns]

### Q9. [Harder] Obtain the first_name, last_name and email address of all employees in the 'Engineering' department who work 0.5 fte_hours or greater.

In [16]:
employees.loc[(employees["department"] == "Engineering")&(employees["fte_hours"] >= 0.5)].loc[:,["first_name", "last_name", "email"]]

first_name   last_name                     email
3      Feodora    Dumingos   fdumingos4@bandcamp.com
37     Sybilla    Lodewick     slodewick12@salon.com
55       Rheba      Booton  rbooton1k@bravesites.com
98      Launce       Feyer                       NaN
99        Shep     Loveday    sloveday2s@twitpic.com
..         ...         ...                       ...
948      Shell        Over           soverqd@icio.us
950     Manuel  Ferrarotti     mferrarottiqf@ovh.net
954      Terry    Sawforde       tsawfordeqj@mit.edu
977        Tam        Tsar        ttsarr6@smh.com.au
984     Pennie    Hanhardt    phanhardtrd@flavors.me

[61 rows x 3 columns]

### Q10. [Harder] Calculate the mean salary of all employees who are members either of the 'Legal' or the 'Accounting' departments.

In [17]:
employees.loc[employees.department.isin(["Legal","Accounting"])].salary.mean()

58213.132530120485

### Q11. [Harder] How many pension enrolled employees are based outside of France, Austria or Ireland?

In [18]:
employees.loc[(employees.pension_enrol == 'Yes')&~(employees.country.isin(["France", "Austria", "Ireland"]))]

id first_name  last_name                      email  \
0       1      Ibbie   Roscrigg       iroscrigg0@google.fr   
2       4     Osmund     Kittel     okittel3@bloomberg.com   
3       5    Feodora   Dumingos    fdumingos4@bandcamp.com   
9      11    Idaline    Pawfoot       ipawfoota@oracle.com   
15     17      Janie  Bourgaize  jbourgaizeg@imageshack.us   
..    ...        ...        ...                        ...   
986   988     Darius       Uman         dumanrf@weebly.com   
992   994       Dani  Berthelet        dbertheletrl@si.edu   
996   998     Wileen     Skones          wskonesrp@hhs.gov   
998  1000  Maribelle      Rotge      mrotgerr@google.co.uk   
999     3     Saleem      Adame         sadame3@utexas.edu   

                   department  team_id  grade         country  fte_hours  \
0                       Legal        9    0.0         Nigeria       0.25   
2                       Legal       10    0.0  United Kingdom       1.00   
3                 Engineering        2    0.0       Indonesia       0.50   
9                   Marketing        1    0.0          Greece       0.25   
15                   Services        6    1.0     Philippines       1.00   
..                        ...      ...    ...             ...        ...   
986      Business Development       10    0.0           China       0.50   
992                     Legal        5    1.0           Chile       1.00   
996                Accounting        4    0.0          Uganda       0.25   
998  Research and Development        5    0.0          Brazil       0.50   
999                  Services        9    0.0        Portugal       0.25   

    pension_enrol   salary  pay_detail_id start_date  
0             Yes  97667.0              1 2014-12-25  
2             Yes  51200.0              4 2007-09-06  
3             Yes  60460.0              5 1990-03-28  
9             Yes  56593.0             11 2017-09-29  
15            Yes  76831.0             17 2005-09-07  
..            ...      ...            ...        ...  
986           Yes  46370.0            988 2006-08-01  
992           Yes  85316.0            994 1999-09-24  
996           Yes  81734.0            998 2000-11-26  
998           Yes  62531.0           1000 2019-03-09  
999           Yes  91749.0              3 1999-10-09  

[473 rows x 13 columns]

### Q12. Add a new column effective_salary to employees containing salary multiplied by fte_hours.

In [19]:
employees["effective_salary"] = employees["salary"] * employees["fte_hours"]

employees

id first_name  last_name                    email  \
0       1      Ibbie   Roscrigg     iroscrigg0@google.fr   
1       2      Sylas  Smallcomb                      NaN   
2       4     Osmund     Kittel   okittel3@bloomberg.com   
3       5    Feodora   Dumingos  fdumingos4@bandcamp.com   
4       6      Peter    de Vaen         pdevaen5@mail.ru   
..    ...        ...        ...                      ...   
995   997    Carlina      Pirot     cpirotro@archive.org   
996   998     Wileen     Skones        wskonesrp@hhs.gov   
997   999      Willy     Dulake  wdulakerq@webeden.co.uk   
998  1000  Maribelle      Rotge    mrotgerr@google.co.uk   
999     3     Saleem      Adame       sadame3@utexas.edu   

                   department  team_id  grade         country  fte_hours  \
0                       Legal        9    0.0         Nigeria       0.25   
1                    Training        3    0.0       Macedonia       0.75   
2                       Legal       10    0.0  United Kingdom       1.00   
3                 Engineering        2    0.0       Indonesia       0.50   
4          Product Management        2    1.0          Greece       0.50   
..                        ...      ...    ...             ...        ...   
995      Business Development        2    0.0           China       1.00   
996                Accounting        4    0.0          Uganda       0.25   
997      Business Development        9    0.0        Colombia       0.25   
998  Research and Development        5    0.0          Brazil       0.50   
999                  Services        9    0.0        Portugal       0.25   

    pension_enrol   salary  pay_detail_id start_date  effective_salary  
0             Yes  97667.0              1 2014-12-25          24416.75  
1              No  48556.0              2 1991-08-01          36417.00  
2             Yes  51200.0              4 2007-09-06          51200.00  
3             Yes  60460.0              5 1990-03-28          30230.00  
4              No  32060.0              6 1992-06-30          16030.00  
..            ...      ...            ...        ...               ...  
995            No  63189.0            997 2018-08-30          63189.00  
996           Yes  81734.0            998 2000-11-26          20433.50  
997            No  21590.0            999 2018-08-01           5397.50  
998           Yes  62531.0           1000 2019-03-09          31265.50  
999           Yes  91749.0              3 1999-10-09          22937.25  

[1000 rows x 14 columns]

### Q13. Obtain a table showing the number of employees in each department.

In [20]:
employees.groupby("department").id.count()

department
Accounting                   72
Business Development         77
Engineering                  87
Human Resources              90
Legal                       102
Marketing                    84
Product Management           79
Research and Development     94
Sales                        80
Services                     73
Support                      81
Training                     81
Name: id, dtype: int64

### Q14. Obtain a count of the number of employees enrolled and not enrolled in the pension scheme. Ignore missing values in pension_enrol for now.

In [21]:
employees.groupby("pension_enrol").id.count()

pension_enrol
No     470
Yes    488
Name: id, dtype: int64

### Q15. [Harder] Repeat your analysis from Question 14 above, but this time fill any missing values in pension_enrol with the string 'Missing'

In [23]:
employees.fillna({'pension_enrol': 'Missing'}).groupby("pension_enrol").id.count()

pension_enrol
Missing     42
No         470
Yes        488
Name: id, dtype: int64